### PR TITLE
Kubelet: implements ImageFsInfo API

### LINKED ISSUE #33048: CRI: Get ImageFS info for the runtime
kubelet currently [relies on cadvisor](https://github.com/kubernetes/kubernetes/blob/12c248315f8191396574085591e1d749ab0fe670/pkg/kubelet/cadvisor/cadvisor_linux.go#L189) to get the ImageFS info for supported runtimes, i.e., docker and rkt. Without this information, the disk manager would not function properly and kubelet will fail to start.

To get the image filesystem information, the following items are needed:
1. Get path from the runtime, i.e., need to add a new call `Info()` to `ImageService` in CRI. 
2. Plumb this information to cadvisor and use it to get `ImageFs` info.

/cc @timstclair @dchen1107 @kubernetes/sig-node

### PR DESCRIPTION
**Pull request by feiskyer**
**What this PR does / why we need it**:

* Implements `ImageFsInfo` for CRI and dockershim
* Kubelet gets image stats from container runtime (instead of cadvisor)
* Follow up of #41733, .

**Which issue this PR fixes** 

closes #33048.

**Special notes for your reviewer**:

**Release note**:

```release-note
CRI: add ImageFsInfo for getting ImageFS info from runtime
```

### CODE REVIEW

File: pkg/kubelet/dockershim/docker_image.go [hunk 1]
```diff
@@ -93,7 +93,9 @@ func (ds *dockerService) RemoveImage(image *runtimeapi.ImageSpec) error {
 	imageInspect, err := ds.client.InspectImageByID(image.Image)
 	if err == nil && imageInspect != nil && len(imageInspect.RepoTags) > 1 {
 		for _, tag := range imageInspect.RepoTags {
-			if _, err := ds.client.RemoveImage(tag, dockertypes.ImageRemoveOptions{PruneChildren: true}); err != nil {
+			if _, err = ds.client.RemoveImage(tag,
+				dockertypes.ImageRemoveOptions{
+					PruneChildren: true}); err != nil {
```
Comment (timstclair) [inline pkg/kubelet/dockershim/docker_image.go]: nit: unnecessary change (go doesn't have a line length limit)

File: pkg/kubelet/disk_manager.go [hunk 2]
```diff
@@ -31,11 +31,14 @@ import (
...
 
+// ImageFsInfoFunc is the interface for getting ImageFsInfo.
+type ImageFsInfoFunc func() (cadvisorapi.FsInfo, error)
+
 // Implementation is thread-safe.
 type diskSpaceManager interface {
 	// Checks the available disk space
 	IsRootDiskSpaceAvailable() (bool, error)
-	IsRuntimeDiskSpaceAvailable() (bool, error)
+	IsRuntimeDiskSpaceAvailable(ImageFsInfoFunc) (bool, error)
```
Comment (timstclair) [inline pkg/kubelet/disk_manager.go]: I think the `ImageFsInfoFunc` should be an implementation detail, not part of the interface.

File: pkg/kubelet/kuberuntime/kuberuntime_image.go [hunk 3]
```diff
@@ -135,14 +137,40 @@ func (m *kubeGenericRuntimeManager) RemoveImage(image kubecontainer.ImageSpec) e
...
-	for _, img := range allImages {
-		stats.TotalStorageBytes += img.Size_
+
+	return &kubecontainer.ImageStats{
+		TotalStorageBytes: imageFSInfo.UsedBytes.GetValue(),
+	}, nil
+}
+
+// ImageFsInfo gets filesytem information of images.
+func (m *kubeGenericRuntimeManager) ImageFsInfo() (cadvisorapiv2.FsInfo, error) {
```
Comment (timstclair) [inline pkg/kubelet/kuberuntime/kuberuntime_image.go]: Why is the `kubeGenericRuntimeManager` constructing a `cadvisorapiv2.FsInfo`? I think the dependencies are a bit scrambled. At the very least, you could clone the API into Kubernetes.

File: cmd/kubelet/app/server.go [hunk 4]
```diff
@@ -966,9 +966,13 @@ func RunDockershim(c *componentconfig.KubeletConfiguration, dockershimRootDir st
 		SupportedPortForwardProtocols:   streaming.DefaultConfig.SupportedPortForwardProtocols,
 	}
 
+	cadvisorInterface, err := cadvisor.New(uint(c.CAdvisorPort), "docker", c.RootDirectory)
```
Comment (timstclair) [inline cmd/kubelet/app/server.go]: I know you never call start on this, but I still don't like the idea of having a second cAdvisor manager in the Kubelet (at the very least, the port should be 0). Since you're only using a single method of the interface, I think it would be better to try and refactor out the piece you need here. I think you should be able to pull out a single static function?

/cc @dashpole

**Review by timstclair — CHANGES_REQUESTED** [review]
This change feels a bit tangled to me, and I didn't trace through everything. I'm hoping that dropping the cAdivsor manager will clean it up. A red-flag is that this relatively small change is touching 32 files...

File: pkg/kubelet/container/runtime.go [hunk 5]
```diff
@@ -157,6 +159,8 @@ type ImageService interface {
 	RemoveImage(image ImageSpec) error
 	// Returns Image statistics.
 	ImageStats() (*ImageStats, error)
+	// ImageFsInfo gets filesytem information of images.
+	ImageFsInfo() (cadvisorapiv2.FsInfo, error)
```
Comment (dashpole) [inline pkg/kubelet/container/runtime.go]: Ideally this API would be `ImageFsInfo() (runtimeapi.FsInfo, error)`, and would replace both ImageStats, and ImageFsInfo here.  cAdvisor wouldnt be part of this api, but just an implementation detail.  I would much rather change the kubelet to use runtimeapi.FsInfo than use a cAdvisor API

File: cmd/kubelet/app/server.go (hunk 4 above)
Comment (feiskyer) [inline cmd/kubelet/app/server.go]: @timstclair They are not running at same time. dockershim mode doesn't start kubelet components.

File: pkg/kubelet/container/runtime.go (hunk 5 above)
Comment (feiskyer) [inline pkg/kubelet/container/runtime.go]: @dashpole Right, this should return `runtimeapi.FsInfo` instead.

**Comment by feiskyer** [comment]
@yujuhong @timstclair @dashpole Addressed comments. PTAL.

File: pkg/kubelet/disk_manager.go [hunk 6]
```diff
@@ -84,7 +91,20 @@ func (dm *realDiskSpaceManager) getFsInfo(fsType string, f func() (cadvisorapi.F
 }
 
 func (dm *realDiskSpaceManager) IsRuntimeDiskSpaceAvailable() (bool, error) {
-	return dm.isSpaceAvailable("runtime", dm.policy.DockerFreeDiskMB, dm.cadvisor.ImagesFsInfo)
+	return dm.isSpaceAvailable("runtime", dm.policy.DockerFreeDiskMB, dm.getRuntimeImageFs)
+}
+
+func (dm *realDiskSpaceManager) getRuntimeImageFs() (cadvisorapiv2.FsInfo, error) {
+	if dm.imageFsInfoFunc == nil {
```
Comment (dashpole) [inline pkg/kubelet/disk_manager.go]: why would this func ever be nil?

Comment (feiskyer) [inline pkg/kubelet/disk_manager.go]: It is nil in all tests.

File: pkg/kubelet/dockershim/remote/docker_service.go [hunk 7]
```diff
@@ -218,5 +217,10 @@ func (d *dockerService) RemoveImage(ctx context.Context, r *runtimeapi.RemoveIma
 
 // ImageFsInfo returns information of the filesystem that is used to store images.
 func (d *dockerService) ImageFsInfo(ctx context.Context, r *runtimeapi.ImageFsInfoRequest) (*runtimeapi.ImageFsInfoResponse, error) {
-	return nil, fmt.Errorf("not implemented")
+	imageFSInfo, err := d.imageService.ImageFsInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	return &runtimeapi.ImageFsInfoResponse{FsInfo: imageFSInfo}, nil
```
Comment (runcom) [inline pkg/kubelet/dockershim/remote/docker_service.go]: This is missing `InodesUsed` from the response as `ImageFsInfo` isn't populating it (and cadvisor doesn't seem to have that field either)

Comment (feiskyer) [inline pkg/kubelet/dockershim/remote/docker_service.go]: Right, so it is left with nil by default.

**Comment by feiskyer** [comment]
ping @timstclair @yujuhong

**Comment by feiskyer** [comment]
@k8s-bot verify test this

File: pkg/kubelet/disk_manager.go (hunk 6 above)
Comment (dashpole) [inline pkg/kubelet/disk_manager.go]: It would be better to define an imageFsInfoFunc for testing, rather than falling back to using cadvisor directly.

Comment (feiskyer) [inline pkg/kubelet/disk_manager.go]: @dashpole Yep, I'd like to do the cleanup in a following PR. WDYT?

**Comment by feiskyer** [comment]
Made another rebase. Ping @yujuhong @timstclair

File: pkg/kubelet/container/helpers.go [hunk 8]
```diff
@@ -365,3 +366,43 @@ func MakePortMappings(container *v1.Container) (ports []PortMapping) {
...
+	}
+	if fsInfo.InodesFree != nil {
+		runtimeFsInfo.InodesAvailable = &runtimeapi.UInt64Value{Value: *fsInfo.InodesFree}
+	}
+
+	return runtimeFsInfo
+}
+
+// ConvertRuntimeFsInfo converts runtimeapi.FsInfo to cadvisorapiv2.FsInfo.
+func ConvertRuntimeFsInfo(imageFsInfo runtimeapi.FsInfo) cadvisorapiv2.FsInfo {
```
Comment (timstclair) [inline pkg/kubelet/container/helpers.go]: I think we should probably convert to runtimeapi.FsInfo everywhere rather than continuing to use the cAdvisor API, which would remove the need for this method. I'm ok with that being a cleanup PR, but maybe leave a TODO here?

File: pkg/kubelet/dockershim/docker_image.go [hunk 9]
```diff
@@ -135,7 +135,30 @@ func getImageRef(client libdocker.Interface, image string) (string, error) {
 
 // ImageFsInfo returns information of the filesystem that is used to store images.
 func (ds *dockerService) ImageFsInfo() (*runtimeapi.FsInfo, error) {
-	return nil, fmt.Errorf("not implemented")
+	imageFSInfo, err := ds.cadvisor.ImagesFsInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	fsInfo := &runtimeapi.FsInfo{
```
Comment (timstclair) [inline pkg/kubelet/dockershim/docker_image.go]: `ConvertCadvisorFsInfo`?

File: pkg/kubelet/kuberuntime/kuberuntime_image.go [hunk 10]
```diff
@@ -135,14 +136,24 @@ func (m *kubeGenericRuntimeManager) RemoveImage(image kubecontainer.ImageSpec) e
 // this is a known issue, and we'll address this by getting imagefs stats directly from CRI.
 // TODO: Get imagefs stats directly from CRI.
 func (m *kubeGenericRuntimeManager) ImageStats() (*kubecontainer.ImageStats, error) {
-	allImages, err := m.imageService.ListImages(nil)
+	imageFSInfo, err := m.imageService.ImageFsInfo()
 	if err != nil {
-		glog.Errorf("ListImages failed: %v", err)
+		glog.Errorf("ImageFsInfo failed: %v", err)
```
Comment (timstclair) [inline pkg/kubelet/kuberuntime/kuberuntime_image.go]: Either return an error or log it, but never both.

File: pkg/kubelet/kuberuntime/kuberuntime_image.go [hunk 11]
```diff
@@ -135,14 +136,24 @@ func (m *kubeGenericRuntimeManager) RemoveImage(image kubecontainer.ImageSpec) e
...
+	return &kubecontainer.ImageStats{
+		TotalStorageBytes: imageFSInfo.UsedBytes.GetValue(),
+	}, nil
+}
+
+// ImageFsInfo gets filesytem information of images.
+func (m *kubeGenericRuntimeManager) ImageFsInfo() (runtimeapi.FsInfo, error) {
+	imageFSInfo, err := m.imageService.ImageFsInfo()
+	if err != nil {
+		glog.Errorf("ImageFsInfo failed: %v", err)
```
Comment (timstclair) [inline pkg/kubelet/kuberuntime/kuberuntime_image.go]: ditto: log or return

File: pkg/kubelet/oom_watcher_test.go [hunk 12]
```diff
@@ -23,7 +23,7 @@ import (
 
 	"k8s.io/client-go/pkg/api/v1"
 	"k8s.io/client-go/tools/record"
-	cadvisortest "k8s.io/kubernetes/pkg/kubelet/cadvisor/testing"
+	cadvisortest "k8s.io/kubernetes/pkg/kubelet/cadvisor/fake"
```
Comment (timstclair) [inline pkg/kubelet/oom_watcher_test.go]: nit: s/cadvisortest/cadvisorfake/

Or just drop the alias. This test is tiny...

File: pkg/kubelet/remote/remote_image.go [hunk 13]
```diff
@@ -130,5 +130,14 @@ func (r *RemoteImageService) RemoveImage(image *runtimeapi.ImageSpec) error {
 
 // ImageFsInfo returns information of the filesystem that is used to store images.
 func (r *RemoteImageService) ImageFsInfo() (*runtimeapi.FsInfo, error) {
-	return nil, fmt.Errorf("not implemented")
+	ctx, cancel := getContextWithTimeout(r.timeout)
+	defer cancel()
+
+	imageFSInfo, err := r.imageClient.ImageFsInfo(ctx, &runtimeapi.ImageFsInfoRequest{})
+	if err != nil {
+		glog.Errorf("ImageFsInfo failed: %v", err)
```
Comment (timstclair) [inline pkg/kubelet/remote/remote_image.go]: ditto: log or return

**Comment by yujuhong** [comment]
I am making some changes to the ImageFs API in #45614. Let's hold on the implementation until the API is revised.

**Comment by feiskyer** [comment]
yep, should hold on for the new stats API.

**Comment by yujuhong** [comment]
@feiskyer are you still working on this PR?

**Comment by feiskyer** [comment]
> @feiskyer are you still working on this PR?

@yujuhong This is depending on #46984 now. It needs the storage identifier from cadvisor.

@WIZARD-CXY claimed to work on it [there](https://github.com/kubernetes/kubernetes/issues/46984#issuecomment-312858590), so I'm still holding on the PR.

**Comment by feiskyer** [comment]
Already added in #51557, closing this one.
